### PR TITLE
expose needs_drop under mem

### PR DESCRIPTION
**Pull request by Gankra**
Note that I purposefully didn't update the intrinsic's documentation, because I think it makes sense for it be more... "honest" about its semantics.

### CODE REVIEW

**Comment by Gankro** [comment]
Fixed the whitespace errors (had auto-whitespace-trimming disabled since Swift people got mad when you fixed that).

The unstable book error thing is uh, new to me.

**Comment by aturon** [comment]
cc @rust-lang/libs @rust-lang/lang

**Comment by alexcrichton** [comment]
👍

File: src/libcore/mem.rs [hunk 1]
```diff
@@ -302,6 +302,58 @@ pub fn align_of_val<T: ?Sized>(val: &T) -> usize {
     unsafe { intrinsics::min_align_of_val(val) }
 }
 
+/// Returns whether dropping values of type `T` matters.
+///
+/// This is purely an optimization hint, and may be implemented conservatively.
+/// For instance, always returning `true` would be a valid implementation of
+/// this function.
```
Comment (eddyb) [inline src/libcore/mem.rs]: So there's some backstory here: basically this was never implemented in the compiler in a principled way, to account for associated types. I've made the `Copy` check recursive recently but I don't believe that's relevant here.
What's left is to normalize associated type projections in the recursion and/or make an auto trait and rely on the trait system to handle the structural recursion involved here. cc @nikomatsakis

**Comment by alexcrichton** [comment]
ping @Gankro looks like this needs a rebase? Looks like there's not a whole lot of opinions here, and we don't have a *huge* amount of ceremony about unstable apis, so I can r+ when you've rebased

**Comment by Mark-Simulacrum** [comment]
@aturon You nominated but didn't label with any teams, I've added both that you pinged (T-lang and T-libs).

**Comment by Gankro** [comment]
Rebased.

File: src/libcore/mem.rs (hunk 1 above)
Comment (gnzlbg) [reply src/libcore/mem.rs]: @eddyb do you think that `needs_drop` can be made accurate ?

**Pull request merged**